ext/test/http/curl_http_test.cc

```diff
@@ -531,3 +532,105 @@ TEST_F(BasicCurlHttpTests, FinishInAsyncCallback)
     }
   }
 }
+
+#ifdef ENABLE_OTLP_COMPRESSION_PREVIEW
+struct GzipEventHandler : public CustomEventHandler
+{
+  ~GzipEventHandler() override = default;
+
+  void OnResponse(http_client::Response & /* response */) noexcept override {}
+
+  void OnEvent(http_client::SessionState state, nostd::string_view reason) noexcept override
+  {
+    is_called_ = true;
+    state_     = state;
+    reason_    = std::string{reason};
+  }
+
+  bool is_called_                  = false;
+  http_client::SessionState state_ = static_cast<http_client::SessionState>(-1);
+  std::string reason_;
+};
+
+TEST_F(BasicCurlHttpTests, GzipCompressibleData)
+{
+  received_requests_.clear();
+  auto session_manager = http_client::HttpClientFactory::Create();
+  EXPECT_TRUE(session_manager != nullptr);
+
+  auto session = session_manager->CreateSession("http://127.0.0.1:19000");
+  auto request = session->CreateRequest();
+  request->SetUri("post/");
+  request->SetMethod(http_client::Method::Post);
+
+  const auto original_size = 1000UL;
+  http_client::Body body(original_size);
+  std::iota(body.begin(), body.end(), 0);
+  request->SetBody(body);
+  request->AddHeader("Content-Type", "text/plain");
+  request->SetCompression(opentelemetry::ext::http::client::Compression::kGzip);
+  auto handler = std::make_shared<GzipEventHandler>();
+  session->SendRequest(handler);
+  ASSERT_TRUE(waitForRequests(30, 1));
+  session->FinishSession();
+  ASSERT_TRUE(handler->is_called_);
+  ASSERT_EQ(handler->state_, http_client::SessionState::Response);
+  ASSERT_TRUE(handler->reason_.empty());
+
+  auto http_request =
+      dynamic_cast<opentelemetry::ext::http::client::curl::Request *>(request.get());
+  ASSERT_TRUE(http_request != nullptr);
+  ASSERT_LT(http_request->body_.size(), original_size);
+
+  session_manager->CancelAllSessions();
+  session_manager->FinishAllSessions();
+}
+
+TEST_F(BasicCurlHttpTests, GzipIncompressibleData)
+{
+  received_requests_.clear();
+  auto session_manager = http_client::HttpClientFactory::Create();
+  EXPECT_TRUE(session_manager != nullptr);
+
+  auto session = session_manager->CreateSession("http://127.0.0.1:19000");
+  auto request = session->CreateRequest();
+  request->SetUri("post/");
+  request->SetMethod(http_client::Method::Post);
+
+  http_client::Body body;
+
+  // Original source:
+  // http://blog.chenshuo.com/2014/05/incompressible-zlibdeflate-data.html
```

Comment:
From a technical point, this test `GzipIncompressibleData` is very good.

The problem here is that the code used below to generate incompressible data comes from this blog, which reads as:

> Copyright 2009 [Shuo Chen's Blog](http://blog.chenshuo.com/) All rights reserved.

aka, not open source.

As a result, and it pains me to say that because this is good work:

Please remove test `GzipIncompressibleData` entirely, unless a different source can be found to generate test data.

Reply:
@chusitoo 
Possible remediation, just define a static array of 256 bytes, with all different values in it, as test data.

Reply:
That's a good catch. Apologies, I should've checked the fine print.

I had already been able to reproduce this using uniform distributions but was more inclined to use the code snippet from that blog post because it was not subject to change from one run to another.

As you rightfully suggested, and for consistency, I think the trade-off could be to generate an incompressible blob once and statically define it in the unit test, instead of relying on the randomly-generated data for every run.